### ISSUE TITLE
Disable Microphone Access Request

### ISSUE THREAD

**Issue by VincentToups** [part-cw/lambdanative#154]
What is the best way to prevent an iOS build from requesting microphone access? 

I'm in the process of iOS App Store review, by the way!

**Comment by mgorges** [comment]
Are you using the [rtaudio](https://github.com/part-cw/lambdanative/tree/master/modules/rtaudio) module? In that case look into `modules/rtaudio/IOS_xml_requireddevicecap` and `modules/rtaudio/IOS_xml_additions` and remove the lines pertaining to the microphone. - All you need to do in the end you need to get the microphone request removed from the `Info.plist`.

**Comment by VincentToups** [comment]
I don't seem to be using the rtaudio module. Here is the contents of my module file:

`config eventloop ln_core ln_glcore ln_glgui audio scran heap pathfinding queue shadchen alist object`

Can I edit the Info.plist manually?

**Comment by VincentToups** [comment]
I've looked at the template file here:

`./loaders/ios/Info.plist.in` and it doesn't seem to have anything like an `NSMicrophoneUsageDescription` key in it. So I am a bit at a loss.

**Comment by mgorges** [comment]
Sorry, I don't think we request microphone access by default, so unless the object module adds it I don't know. You don't want to look for `NSMicrophoneUsageDescription` but rather `<string>microphone</string>` in a `IOS_xml_requireddevicecap` or maybe a background mode that requires it?

**Comment by mgorges** [comment]
Can you attach the Info.plist generated by your app as a gist, please? You find it in the `~/Library/Caches/lambdanative/build/APPNAME` folder.

**Comment by VincentToups** [comment]
Here it is. I don't see anything about microphones.

https://gist.github.com/VincentToups/ce348e171e2a00ccf37e14b4f5c84439

**Comment by mgorges** [comment]
Okay, so why do you think it requests microphone access then - did it pop up a box on application start asking about it?

**Comment by VincentToups** [comment]
It pops up a request when started on the iPhone. I believe that it just does it once, after you accept, so I had forgotten about it. But the iOS App Store Review caught it. 
<img width="502" alt="screen shot 2017-05-30 at 8 45 24 pm" src="https://cloud.githubusercontent.com/assets/31994/26610971/f25ec9de-4578-11e7-8ff2-9f028772bcca.png">

**Comment by VincentToups** [comment]
I have already tried a make clean. What is the heavy duty clean called again, "scrub?"

**Comment by mgorges** [comment]
Another potential thing to look at is [loaders/ios/mo_audio.mm#L515](https://github.com/part-cw/lambdanative/blob/master/loaders/ios/mo_audio.mm#L515), which requests `kAudioSessionCategory_PlayAndRecord`. A potential alternative might be `kAudioSessionCategory_MediaPlayback` or `kAudioSessionCategory_SoloAmbientSound`

**Comment by mgorges** [comment]
And yes `make scrub` rebuilds everything but it shouldn't be needed in this case

**Comment by VincentToups** [comment]
This diff seems to have resolved the issue:

```
diff --git a/loaders/ios/mo_audio.mm b/loaders/ios/mo_audio.mm
index b78442b..d7482fd 100644
--- a/loaders/ios/mo_audio.mm
+++ b/loaders/ios/mo_audio.mm
@@ -512,7 +512,7 @@ bool MoAudio::init( Float64 srate, UInt32 frameSize, UInt32 numChannels )
         return false;
     }
 
-    UInt32 category = kAudioSessionCategory_PlayAndRecord;
+    UInt32 category = kAudioSessionCategory_MediaPlayback;
     // set audio category
     err = AudioSessionSetProperty( kAudioSessionProperty_AudioCategory, sizeof(category), &category );
     if( err )
```

**Comment by VincentToups** [comment]
Thank you!

**Comment by VincentToups** [comment]
Unfortunately, it seems like my iPhone is simply remembering my answer to the previous build of the game  - during testing for the approval, it is still requesting microphone access.

Any ideas?

**Comment by mgorges** [comment]
@VincentToups sorry not on top of my head - might be a few days to sort out.

**Comment by mgorges** [comment]
to help with finding the issue, if you don't include any audio module, does an app you create still prompt for microphone access on its first run?

**Comment by VincentToups** [comment]
It seems like that patch I posted above did resolve the issue, but it took
a bit of doing to get the new version percolated through testing. I'm not
sure what the broader implications of the change above are, but it does
resolve the issue.

**Comment by VincentToups** [comment]
Hate to churn this up again, but after further testing I have realized that while changing the session category to MediaPlayback does prevent the app from requesting Microphone access, it also disables all audio.

I've tried other audio session categories (Ambient, SoloAmbient, Default, etc) and none seem to work except PlayAndRecord, which does, unfortunately, continue to request microphone access. 

I'm in the process of trying to understand enough of the iOS developer docs to resolve the issue (I have a hunch it has something to do with the logic by which the app responds to session interruptions, since it seems like non-PlayAndRecord categories _also_ cause a crash when the app is refocused) but if you could share any knowledge you have, that might help me along.

**Comment by VincentToups** [comment]
As I try to resolve the issue with using categories other than PlaybackAndRecord, I am trying to instrument the mo_audio.mm file with calls to NSLog, for instance:

```
NSLog(@"Lambdanative - in audio propListener.");
```

These don't appear to show up in the console, however, even when I compile in debug mode. Does lambdanative do something to disable NSLog calls?

**Comment by mgorges** [comment]
@VincentToups it shouldn't and if you search through the code we use those calls in a few places. I am wondering if the build process, which needs to make both 32bit and 64bit apps and then merge them together doesn't rebuild the mo_audio.mm properly, as it is not a file we test for. Maybe touch a file like `modules/audio/audiofile.scm` to see if that forces the rebuild?

**Comment by VincentToups** [comment]
I can tell that the module is being rebuilt because if I change the kAudioSessionCategory to MediaPlayback, for instance, the sound stops working and I stopped being queried for Microphone access. Or, if I add a syntax error or undefined identifier, the build fails.

**Comment by VincentToups** [comment]
So according to this: 

https://github.com/alexbw/novocaine/issues/76

Commenting lines 

```
    /* UInt32 one = 1; */
    /* // enable input */
    /* err = AudioUnitSetProperty( inRemoteIOUnit, kAudioOutputUnitProperty_EnableIO,  */
    /*                             kAudioUnitScope_Input, 1, &one, sizeof(one) ); */
    /* if( err ) */
    /* { */
    /*     // TODO: "couldn't enable input on the remote I/O unit" */
    /*     return false; */
    /* } */
```

In a similar project did the trick in a similar situation. However, for me, these lines (362-371), when commented out, also appear to prevent the audio from working. I've tried commenting all the similar expressions that seem to pertain to input channel or microphone, but no luck so far.

**Comment by VincentToups** [comment]
(Note that this seems to be using novocaine, which I notice is also being built with corpse wizard, though I don't believe its being used).

**Comment by mgorges** [comment]
Regarding the NSLog, Apple broke that in the combination of iOS 10 and Xcode 8. https://stackoverflow.com/questions/39584707/nslog-on-devices-in-ios-10-xcode-8-seems-to-truncate-why has a potential fix, but I am looking for an BSD licensed piece of code to include

**Comment by mgorges** [comment]
Should be fixed in 5e7849f. Please test and let me know.

**Comment by VincentToups** [comment]
This seems to do the trick - thank you!
